### DISCUSSION TITLE
How do I include hidden directories but exclude .git folders?

### DISCUSSION THREAD

**Discussion by haji-ali** [BurntSushi/ripgrep#1578] [General] [answered]
Is there a way to include hidden files/folders but exclude `.git` folders? I tried

```
rg --hidden -g '!*.git*' string
```

But this would exclude all files that have the word `.git` in them, including files that I want to include like `.gitignore` or if I ever have a file like `my.git`. Since `-g` matches a glob I cannot match path separators. Is there a way to achieve this?

**Comment by okdana** [comment, marked as answer]
The glob patterns are meant to work like the ones in [`.gitignore`](https://git-scm.com/docs/gitignore#_pattern_format); you can just use `!.git` (or `!.git/` to be more precise)

```
% git init
Initialized empty Git repository in /home/dana/test/.git/
% \rg --no-ignore -q --hidden refs && echo match || echo no match
match
% \rg --no-ignore -q --hidden -g '!.git/' refs && echo match || echo no match
no match
```

**Comment by BurntSushi** [comment]
Also note that you can put `.git/` in your `.ignore` (or `.rgignore`), which will have the same effect if you don't want to specify `-g '!.git/'` every time you use `--hidden`.

**Comment by garrison** [comment]
In order to get this to be the default behavior, I needed the following two lines in ripgreprc:
```
--hidden
--glob=!.git/
```
If the second line is instead `-g !.git/`, it does not behave properly.

**Reply by BurntSushi** [reply]
You'll want to read the docs for the config file format carefully. Every line is a single argument. So you need either

```
--glob=!.git/
```

or

```
--glob
!.git/
```

**Comment by 3052** [comment]
the existing answers seem pointlessly complicated. I am sure someone will correct me, but this seems to work for my needs:

~~~
rg -u -u -g !.git SOMETHING
~~~

**Reply by BurntSushi** [reply]
That's basically the same as what folks have suggested, except you're missing the single quotes around the `!.git` and you're also ignoring all other gitignores.

**Reply by BurntSushi** [reply]
Yes, but that's not obvious to anyone else following along. Typing an unquoted `!` in a Unix shell will get you odd results:

```
$ rg -u -u -g !.git SOMETHING
zsh: event not found: .git
```

>  and sometimes I want to search EVERYTHING except .git

Valid, but not what this question asked. It's fine to share it here, but it would be good to note the differences between what you're answering with and what the OP asked for.

**Reply by 3052** [reply]
I mean honestly the original question and my variant seem VERY common use case, so if a FAQ or similar exists this should be on it

**Reply by BurntSushi** [reply]
I think there's some miscommunication here. I'm not saying it shouldn't be in a FAQ or is uncommon.

**Reply by BurntSushi** [reply]
> right, Ive already read and responded to your points, so I was raising a new point

If you want to "raise a new point," then open a new question or issue.

**Reply by 3052** [reply]
no thanks, I am happy with it above as a comment
